### PR TITLE
[REVIEW] Add maximum_pool_size parameter to reinitialize API

### LINKED ISSUE #519: [FEA] add arg to Python API reinitialize() to set max pool size (and enforce usage when managed_memory=True?)
Because the default maximum pool size is all memory for the device, setting `managed_memory=True` in the RMM Python `reinitialize()` API still does not allow for oversubscribing GPU memory.  This makes sense (what would the new default max pool size be?) but is still unexpected, since in most cases the caller is expecting UVM features like oversubscribing when setting `managed_memory=True`, and instead that aspect is silently ignored by default.

This issue proposes an additional argument to `rmm.reinitialize()` to allow the caller to set a new maximum pool size from the same call that allows them to enable managed_memory in the first place.

In addition to the new arg, it might make sense to enforce that setting a new max pool size is done when enabling `managed_memory` for the reason mentioned above (ie. enabling UVM implies the ability to oversubscribe, so make the user aware that a new max pool size would be needed). However, for the case of a user enabling UVM in order to have a single address space across devices (is this done from Python?), then perhaps forcing them to also set a new max pool size might not be the best UX.

### PR DESCRIPTION
**Pull request by shwina**
Closes #519 

cc: @rlratzel

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/rmm/rmm.py [hunk 1]
```diff
@@ -50,8 +51,10 @@ def reinitialize(
         If True, use managed memory for device memory allocation
     initial_pool_size : int, default None
         When `pool_allocator` is True, this indicates the initial pool size in
-        bytes. None is used to indicate the default size of the underlying
-        memorypool implementation, which currently is 1/2 total GPU memory.
+        bytes. By default, 1/2 of the total GPU memory is used.
+    maximum_pool_size : int, default None
+        When `pool_allocator` is True, this indicates the maximum pool size in
```
Comment (rlratzel) [inline python/rmm/rmm.py:56]: This is a bit pedantic, but what's the behavior when `maximum_pool_size` is set when `pool_allocator` is False? I'm assuming it's ignored?

File: python/rmm/tests/test_rmm.py [hunk 2]
```diff
@@ -352,3 +352,20 @@ def test_binning_memory_resource(dtype, nelem, alloc, upstream_mr):
     assert rmm.mr.get_current_device_resource_type() is type(mr)
     array_tester(dtype, nelem, alloc)
     rmm.reinitialize()
+
+
+def test_reinitialize_max_pool_size():
```
Comment (rlratzel) [inline python/rmm/tests/test_rmm.py:357]: Should there also be a test for a user setting the `initial_pool_size` > `maximum_pool_size`?  I'm assuming that results in `MemoryError`?

File: python/rmm/_lib/memory_resource.pyx [hunk 3]
```diff
@@ -312,11 +313,12 @@ cpdef void _initialize(
...
-        if initial_pool_size is None:
-            args = (upstream(),)
-        else:
-            args = (upstream(), initial_pool_size)
-
+        args = [upstream()]
+        if initial_pool_size is not None:
+            args.append(initial_pool_size)
+        if maximum_pool_size is not None:
+            args.append(maximum_pool_size)
```
Comment (kkraus14) [inline python/rmm/_lib/memory_resource.pyx:320]: This looks fishy, if depending on the value of `initial_pool_size`, the argument for `maximum_pool_size` can be at position 1 or 2 in the list.

File: python/rmm/tests/test_rmm.py (hunk 2 above)
Comment (shwina) [reply python/rmm/tests/test_rmm.py:357]: That's handled by libcudf and it does raise an error. Will add a test for it.

File: python/rmm/_lib/memory_resource.pyx (hunk 3 above)
Comment (shwina) [reply python/rmm/_lib/memory_resource.pyx:320]: Ah nice catch - oversight on my part

File: python/rmm/rmm.py (hunk 1 above)
Comment (shwina) [reply python/rmm/rmm.py:56]: Yes - it is ignored

Comment (kkraus14) [reply python/rmm/rmm.py:56]: Could we indicate such in the docstring, and possibly considering throwing a Warning or Exception if someone uses it in an ignored context?

File: python/rmm/_lib/memory_resource.pyx [hunk 4]
```diff
@@ -49,22 +49,30 @@ cdef class PoolMemoryResource(MemoryResource):
...
+            maximum_pool_size=None
     ):
+        cdef size_t c_initial_pool_size
+        cdef size_t c_maximum_pool_size
+        c_initial_pool_size = (
+            ~0 if initial_pool_size is None else initial_pool_size
+        )
+        c_maximum_pool_size = (
+            ~0 if maximum_pool_size is None else maximum_pool_size
+        )
```
Comment (shwina) [inline python/rmm/_lib/memory_resource.pyx:62]: AFAICT there's no super clean way to handle this (e.g., the way @harrism would like to see as detailed in https://github.com/rapidsai/rmm/issues/497). Python needs to be able to handle when the user specifies the `maximum_pool_size` but not the `initial_pool_size`. There's no way to do that in C++ without explicitly specifying the sentinel value `~0` for `initial_pool_size` (is there?).

File: python/rmm/rmm.py (hunk 1 above)
Comment (shwina) [reply python/rmm/rmm.py:57]: I made it explicit in the docstring, but I'm on the fence about throwing/warning. Could be annoying for someone wanting to quickly switch between `pool=True/False`

File: python/rmm/tests/test_rmm.py (hunk 2 above)
Comment (shwina) [reply python/rmm/tests/test_rmm.py:357]: Added another test

File: python/rmm/_lib/memory_resource.pyx (hunk 4 above)
Comment (kkraus14) [reply python/rmm/_lib/memory_resource.pyx:62]: I think we'd need another C++ overload or to change the C++ API to have a builder pattern or something similar. cc @harrism @jrhemstad

Comment (kkraus14) [reply python/rmm/_lib/memory_resource.pyx:62]: Merging as the Python side is fine for now. We can move to a different strategy if desired in a follow up PR.

Comment (harrism) [reply python/rmm/_lib/memory_resource.pyx:62]: @shwina you should access the default value from the C++ class.

**Pull request merged**